src/server/dflycmd.cc

```diff
@@ -513,13 +506,27 @@ void DflyCmd::TakeOver(CmdArgList args, RedisReplyBuilder* rb, ConnectionContext
 
   atomic_bool catchup_success = true;
   if (*status == OpStatus::OK) {
-    dfly::SharedLock lk{replica_ptr->shared_mu};
-    auto cb = [replica_ptr = replica_ptr, end_time, &catchup_success](EngineShard* shard) {
-      if (!WaitReplicaFlowToCatchup(end_time, replica_ptr.get(), shard)) {
-        catchup_success.store(false);
-      }
-    };
-    shard_set->RunBlockingInParallel(std::move(cb));
+    // We need to loop or otherwise we can't guarantee that there is no data loss.
+    // *All* nodes must be in sync when the takeover completes such that when the replica
+    // reconciles with the new promoted master the only data that needs to be sent is the
+    // journaled commands that came after the takeover. On the contrary, If nodes were not
+    // in sync after the takeover, then any data that was not replicated from the old master
+    // is lost, because the new master does not contain their entries and a fallback to FULL
+    // SYNC is required (since the journal starts when the takeover completes, which implies
+    // data parity with the old master and an empty journal on the newly promoted master).
+    util::fb2::LockGuard mu_lk(mu_);
+    for (auto [id, repl_ptr] : replica_infos_) {
+      dfly::SharedLock lk{repl_ptr->shared_mu};
+      // Wait for ALL replicas to catch up. Otherwise we can't do partial sync from same source
+      // master. If the replica misses 1 LSN entry, that entry will never exist in the LSN buffer
+      // of the new master effectively killing partial sync.
```

Comment:
```suggestion

      // Wait for ALL replicas to catch up. Otherwise we can't do partial sync from same source
      // master. If the other replica misses a single journal entry, that entry will never exist in the replication buffer
      // of the new master, so that replica won't be able to use partial sync to sync with the new master.
```